app/src/main/java/com/github/multimatum_team/multimatum/viewmodel/DeadlineListViewModel.kt

```diff
@@ -58,32 +59,38 @@ class DeadlineListViewModel @Inject constructor(
     /**
      * Re-fetch all deadlines from the repository and assign the value to the LiveData.
      */
-    private fun refreshDeadlines(callback: (Map<DeadlineID, Deadline>) -> Unit) = viewModelScope.launch {
-        _deadlines.value = deadlineRepository.fetchAll()
-        callback(_deadlines.value!!)
-    }
+    private fun refreshDeadlines(callback: (Map<DeadlineID, Deadline>) -> Unit = {}) =
+        viewModelScope.launch {
+            _deadlines.value = deadlineRepository.fetchAll()
+            callback(_deadlines.value!!)
+        }
 
     /**
      * Get all deadlines.
      */
     fun getDeadlines(): LiveData<Map<DeadlineID, Deadline>> =
         _deadlines
 
+    fun getDeadline(id: DeadlineID): Deadline =
+        _deadlines.value!![id]!!
+
     /**
      * Add a new deadline to the repository.
      */
-    fun addDeadline(deadline: Deadline, callback: (DeadlineID)->Unit) = viewModelScope.launch {
-        val id = deadlineRepository.put(deadline)
-        callback(id)
-    }
+    fun addDeadline(deadline: Deadline, callback: (DeadlineID) -> Unit = {}) =
+        viewModelScope.launch {
+            val id = deadlineRepository.put(deadline)
+            callback(id)
+        }
```

Comment:
If I understand correctly, the goal of the callback argument is to be able to execute something in the `launch` block, but why is that needed? In what way is that different from simply returning the id and executing the callback after `addDeadline` returned?

Reply:
We cannot return from a `Job` (`viewModelScope.launch` returns a `Job` instance). `launch` is some kind of `fork` operation which returns immediately, it starts a task on the viewmodel thread and returns a reference to the job so that we can chose to cancel it for example, but it doesn't wait for the job to complete before returning. Also, unlike coroutines and `CompletableFuture`s, the `Job` class does not provide an equivalent to `addOnCompleteListener` so we have to handle callbacks ourselves.